src/library/crate/cratefeature.cpp

```diff
@@ -0,0 +1,915 @@
+#include "library/crate/cratefeature.h"
+
+#include <QFileDialog>
+#include <QInputDialog>
+#include <QLineEdit>
+#include <QMenu>
+#include <QDesktopServices>
+
+#include <algorithm>
+
+#include "library/export/trackexportwizard.h"
+#include "library/library.h"
+#include "library/parser.h"
+#include "library/parserm3u.h"
+#include "library/parserpls.h"
+#include "library/parsercsv.h"
+#include "library/trackcollection.h"
+#include "library/treeitem.h"
+
+#include "sources/soundsourceproxy.h"
+
+#include "widget/wlibrary.h"
+#include "widget/wlibrarytextbrowser.h"
+
+#include "util/dnd.h"
+
+
+namespace {
+
+QString formatLabel(
+        const CrateSummary& crateSummary) {
+    return QString("%1 (%2) %3").arg(
+        crateSummary.getName(),
+        QString::number(crateSummary.getTrackCount()),
+        crateSummary.getTrackDurationText());
+}
+
+} // anonymous namespace
+
+CrateFeature::CrateFeature(Library* pLibrary,
+                           TrackCollection* pTrackCollection,
+                           UserSettingsPointer pConfig)
+        : LibraryFeature(pConfig),
+          m_cratesIcon(":/images/library/ic_library_crates.png"),
+          m_lockedCrateIcon(":/images/library/ic_library_locked.png"),
+          m_pTrackCollection(pTrackCollection),
+          m_crateTableModel(this, pTrackCollection) {
+
+    initActions();
+
+    // construct child model
+    m_childModel.setRootItem(std::make_unique<TreeItem>(this));
+    rebuildChildModel();
+
+    connectLibrary(pLibrary);
+    connectTrackCollection();
+}
+
+CrateFeature::~CrateFeature() {
+}
+
+void CrateFeature::initActions() {
+    m_pCreateCrateAction = std::make_unique<QAction>(tr("Create New Crate"),this);
```

Comment:
Now the QAction has two owners.This should become 
```m_pCreateCrateAction = make_parented_raw<QAction>(tr("Create New Crate"),this);```
after https://github.com/mixxxdj/mixxx/pull/1161 is merged